src/GraphBLAS-sharp.Backend/Matrix/CSRMatrix/Map.fs

```diff
@@ -127,3 +127,134 @@ module internal Map =
         fun (queue: MailboxProcessor<_>) allocationMode (matrix: ClMatrix.CSR<'a>) ->
             mapToCOO queue allocationMode matrix
             |> toCSRInplace queue allocationMode
+
+    module WithValue =
+        let preparePositions<'a, 'b, 'c> (clContext: ClContext) workGroupSize op =
+
+            let preparePositions (op: Expr<'a option -> 'b option -> 'c option>) =
+                <@ fun (ndRange: Range1D) (operand: ClCell<'a option>) rowCount columnCount (values: ClArray<'b>) (rowPointers: ClArray<int>) (columns: ClArray<int>) (resultBitmap: ClArray<int>) (resultValues: ClArray<'c>) (resultRows: ClArray<int>) (resultColumns: ClArray<int>) ->
+
+                    let gid = ndRange.GlobalID0
+
+                    if gid < rowCount * columnCount then
+
+                        let columnIndex = gid % columnCount
+                        let rowIndex = gid / columnCount
+
+                        let startIndex = rowPointers.[rowIndex]
+                        let lastIndex = rowPointers.[rowIndex + 1] - 1
+
+                        let value =
+                            (%Search.Bin.inRange) startIndex lastIndex columnIndex columns values
+
+                        match (%op) operand.Value value with
+                        | Some resultValue ->
+                            resultValues.[gid] <- resultValue
+                            resultRows.[gid] <- rowIndex
+                            resultColumns.[gid] <- columnIndex
+
+                            resultBitmap.[gid] <- 1
+                        | None -> resultBitmap.[gid] <- 0 @>
+
+            let kernel = clContext.Compile <| preparePositions op
+
+            fun (processor: MailboxProcessor<_>) (operand: ClCell<'a option>) rowCount columnCount (values: ClArray<'b>) (rowPointers: ClArray<int>) (columns: ClArray<int>) ->
```

Comment:
Is it necessary to pass all the fields of the matrix one by one? Is it possible to pass the entire matrix at once?